### ISSUE TITLE
Solution for https and reverse proxy - WORKS

### ISSUE THREAD

**Issue by markosole** [rikulo/socket.io-client-dart#319]
Hi all,

After 3 days of searching and digging around, I've finally made comatible setup between NodeJS Server + Client + Flutter client that work over HTTPS. Many of you had or still have same problem and here is short and sweet how to set it up with Apache reverse proxy forwarding

# Server
Short and simple
```javascript
const express = require('express');
const app = express();
const http = require("http").createServer(app);
socketio = require("socket.io")(http);

http.listen(5003, () => console.log('Socket server served on port: 5003'));
```

# NodeJS Client
Works with https with transports set to `polling` and without specified port number. 
Sidenote: NodeJS client does not work with `websocket` transports. It has to be set to "polling" mode or not set at all (default is polling)

```javascript

var ios = require('socket.io-client');
var socketClient = ios.connect('https://domain.com', {
    reconnect: true,
    secure: true,
    rejectUnauthorized : false,
    transports: ['polling']

});
```

# Flutter Client
Works with **http** and on specified port:

```dart
IO.Socket socket = IO.io('http://domain.com:5003',  
    OptionBuilder()  
        .setTransports(['websocket'])  
        .build());  
  
socket.connect();
```

Works with HTTPS:

```dart
IO.Socket socket = IO.io('https://domain.com',  
    OptionBuilder()  
        .setTransports(['websocket'])  
        .build());  
  
socket.connect();
```

## Apache config - common setups
This does not work for Flutter client on https. Use another one below.

```c
    ProxyPass / http://127.0.0.1/
    RewriteEngine on
    RewriteCond %{HTTPS:Upgrade} =websocket
    RewriteRule /(.*) "http://127.0.0.1:5003/$1" [P,L]
    RewriteCond %{HTTP:Upgrade} !=websocket
    RewriteRule /(.*) "http://127.0.0.1:5003/$1" [P,L]
```

## Apache config - Flutter compatible
Keep in mind to add ` ws:// ` instead ` http:// ` as on example above

```c
     RewriteEngine On
    #RewriteCond %{REQUEST_URI}  ^/socket.io            [NC]
    #RewriteCond %{QUERY_STRING} transport=websocket    [NC]
    
    # Needed for Flutter app
    RewriteRule /(.*)           ws://localhost:5003/$1 [P,L]

    # Needed for NodeJS clients
    ProxyPass / http://localhost:5003/
    ProxyPassReverse / http://localhost:5003/
```

# Versions used
NodeJS socket.io server: 4.6.0
NodeJS socket.io-client: 4.6.0
Flutter socket_io_client package: 2.0.1

Drop a comment if you have questions and suggestions. Cheers

**Comment by sametserpil** [comment]
@markosole Hello. How can I edit rules for Nginx? I can connect my remote server socket via javascript client but not via Flutter.

**Comment by markosole** [comment]
@sametserpil is your server Apache or Nginx and do you have access to config? - I am not Apache (and even less Nginx) expert but I'll try to help.

**Comment by sametserpil** [comment]
Yeah server is mine and I've setup Nginx for reverse proxy. Here are my logs: 

opening https://my-host-url-here
I/flutter ( 9768): FINE: 2023-03-07 14:47:11.806365: creating transport "polling"
I/flutter ( 9768): FINE: 2023-03-07 14:47:11.813039: setting transport websocket
I/flutter ( 9768): FINE: 2023-03-07 14:47:11.814809: connect attempt will timeout after 20000
I/flutter ( 9768): FINE: 2023-03-07 14:47:11.956447: socket close with reason: "transport close"

**Comment by markosole** [comment]
Can you paste your:

1. Ngix configuration
2. Flutter socket connection part
3. Node JS server / Client connection part of the code

It's missconfiguration or missmatch of versions.

**Comment by sametserpil** [comment]
My Nginx conf:

```
location / {
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
}
```

My versions are as follows:

NodeJS -> v18.14.1
Socket.io Server -> "socket.io": "^4.6.1"
Socket.io Android Java Client -> io.socket:socket.io-client:2.1.0 (This is working)
Socket.io NodeJS client -> "socket.io-client": "^4.6.1" (This is also working)
Socket.io Flutter client -> socket_io_client: ^2.0.1 (This is not working!)

Here how I try to connect:
```
IO.Socket socket = IO.io('https://my-url-here');
    socket.onConnect((_) {
      print('connect');
      socket.emit('msg', 'test');
    });
```

My NodeJS code:
```
const server = require('http').createServer();
const io = require('socket.io')(server);
io.on('connection', client => {
    console.log('connected'+client.id);
    client.on('event', data => { /!* … *!/ });
    client.on('disconnect', () => { /!* … *!/ });
});
server.listen(3000);
```

**Comment by Jimesh843** [comment]
Connection Error: WebSocketException: Connection to 'https://example.com:0/socket.io/?EIO=4&transport=polling#' was not upgraded to websocket 

I am facing same issue. I have used socket_io_client: ^2.0.3+1 and from backend i have used 4.7.3 version of node and i have also set code necessary for server. but still can't connect

**Comment by jumperchen** [comment]
@Jimesh843 
`transport=polling` only supports under web environment.

**Comment by Jimesh843** [comment]
then i need to use transport as websocket instaed of pooling?

**Comment by Jimesh843** [comment]
> then i need to use transport as websocket instaed of pooling?
> 
> socket = IO.io( 'https://example.com', <String, dynamic>{ 'autoConnect' : false, 'transports' : ['websocket'], 'extraHeaders': {'foo': 'bar'}, }, );
> 
> is it relevant?
> 
> Output: Connection Error: WebSocketException: Connection to 'https://example.com:0/socket.io/?EIO=4&transport=websocket#' was not upgraded to websocket
> 
> Still I am getting this error should i need to give any thing on server side or any permission from my side

Please help to solve this issue

**Comment by devilaark2018** [comment]
We are trying it for our Game Project, the adding of ":0" at the end of the domain is causing issue even i add :80 it don't work it expect the port of at least 4 digits.
The same code work on flutter web but not on android device
